### ISSUE TITLE
What to do about nvim not recognizing aliases/symlinks from vimdiff?

### ISSUE THREAD

**Issue by RichardBronosky** [neovim/neovim#14035]
### Actual behaviour

Before reporting this as a bug, I found [vim_diff.html#nvim-features-removed][nvim-features-removed]  which explains that you know that this vim feature doesn't work in nvim. But, you do not explain why. Also, in [nvim.html#nvim-from-vim][nvim-from-vim] there are no matches for `vimdiff`. Finally, in [diff.html#diff-mode][diff-mode] there are no matches for `link` or `alias`.

I can think of at least 4 ways to address this, but I don't know why the decision was made. That affects how I solve it (and what the impact will be on other processes that expect to find `vimdiff` in my `$PATH`).

### Expected behaviour

My expectation (and proposed solution) is:
1. [vim_diff.html][nvim-features-removed]
    1. Explain why "These Vim features were intentionally removed from Nvim"
    1. At least for `vimdiff`
1. [nvim.html][nvim-from-vim]
    1. Explain what to do about the `vimdiff` alias being removed
1. [diff.html][diff-mode]
    1. Link to the explanation of _what to do about the `vimdiff` alias being removed_

If someone will help me with _1.ii_, I will submit a PR for all of these. As a vim user since 1998, it's really important that some basic expectations "just work" in order for me to transition.

- `nvim --version`:
```
NVIM v0.4.4
Build type: Release
LuaJIT 2.1.0-beta3
Compilation: clang -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -DNDEBUG -DMIN_LOG_LEVEL=3 -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wmissing-prototypes -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fno-common -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -I/tmp/neovim-20210221-16038-9kge3w/neovim-0.4.4/build/config -I/tmp/neovim-20210221-16038-9kge3w/neovim-0.4.4/src -I/usr/local/include -I/usr/local/opt/luv/include -I/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include -I/usr/local/opt/gettext/include -I/tmp/neovim-20210221-16038-9kge3w/neovim-0.4.4/build/src/nvim/auto -I/tmp/neovim-20210221-16038-9kge3w/neovim-0.4.4/build/include
Compiled by brew@iMac-Pro

Features: +acl +iconv +tui
```

- Operating system/version:
```
macOS 10.15.7
```

- Terminal name/version:
```
alacritty 0.7.1 (a852eb8)
```

- `$TERM`:
```
screen-256color
```

[nvim-features-removed]: https://neovim.io/doc/user/vim_diff.html#nvim-features-removed
[nvim-from-vim]: https://neovim.io/doc/user/nvim.html#nvim-from-vim
[diff-mode]: https://neovim.io/doc/user/diff.html#diff-mode

**Comment by mjlbach** [comment]
I don't understand, is the issue with the documentation or the use of `nvim -d` for the diff?

**Comment by jdhao** [comment]
@RichardBronosky  If your aim happen to be using nvim as diff tool for git, I think the following works fine:

```
[diff]
    tool = nvimdiff
[difftool]
    prompt = false
[difftool "nvimdiff"]
    cmd = "nvim -d \"$LOCAL\" \"$REMOTE\""
[merge]
    tool = nvimdiff
[mergetool]
    prompt = true
[mergetool "nvimdiff"]
    cmd = "nvim -d \"$LOCAL\" \"$REMOTE\" \"$MERGED\" -c 'wincmd w' -c 'wincmd J'"
```

**Comment by RichardBronosky** [comment]
It's both. But the "egg" is **knowing why** it was decided to not have `ln -s /usr/local/bin/nvim /usr/local/bin/vimdiff` cause nvim to react to being called as `vimdiff` have the same result as `nvim -d`. If I know that, I can know how to satisfy the need holistically (not just for the shell or for git). Then I can document it and submit a PR.

**Comment by mjlbach** [comment]
If it bothers you, you can add this to your shell.

```bash
vimdiff() {
  nvim -d "$@"
}
```

Maybe someone else can provide the historical context that informs neovim's command-line arguments.

**Comment by jdhao** [comment]
@RichardBronosky  My 2 cents on this: the Vim package will install `vim`, `vimdiff` and `vimtutor` etc. into your system, but these aliases can be trivially implemented using `nvim` plus some command (`vimtutor` is just `nvim -c "Tutor"`, e.g.). Do we really need these clumsy aliases?  Maybe they are removed just for simplicity's sake 😃
I think @justinmk  may know the rationale behind this.

**Comment by clason** [comment]
Also, to make it possible to have neovim and vim installed in parallel (for testing, if nothing else). "Stealing" vim's alias would not be playing nice; neovim is not a hostile takeover.

**Comment by RichardBronosky** [comment]
This is why it's [actually] complex. While the proposed solutions:
- Update ~/.gitconfig
- Add a shell function
- Add a shell alias
all satisfy a subset of users' needs, they are not holistic in the way that `ln -s $(dirname $(which vim))/vimdiff $(which vim)` is. As a symlink [presumably] in your $PATH, it is accessible to anything that expects to find `vimdiff`. Putting an alias/function in your shell's **rc** file vs. **profile** file [is a complex subject](https://www.golinuxcloud.com/bashrc-vs-bash-profile/). However, even that doesn't work for running `vimdiff` from a process other than a shell. For example, a remote command via `ssh`. See:
```
$ man ssh | awk '/SYNOPSIS|If a command|PATH|^ENVIRONMENT/{p=1; print ""} /^$/ && p==1{p=0} p==1{print}'

SYNOPSIS
     ssh [-46AaCfGgKkMNnqsTtVvXxYy] [-B bind_interface] [-b bind_address] [-c cipher_spec]
         [-D [bind_address:]port] [-E log_file] [-e escape_char] [-F configfile] [-I pkcs11]
         [-i identity_file] [-J destination] [-L address] [-l login_name] [-m mac_spec]
         [-O ctl_cmd] [-o option] [-p port] [-Q query_option] [-R address] [-S ctl_path]
         [-W host:port] [-w local_tun[:remote_tun]] destination [command]

     If a command is specified, it is executed on the remote host instead of a login shell.

ENVIRONMENT
     ssh will normally set the following environment variables:

     PATH                  Set to the default PATH, as specified when compiling ssh.
```

What that tells you is that nothing proposed would work. Therefore, the only viable alternative to a symlink is a wrapper script. I have no problem doing that. I still don't know why it was decided that `nvim` should not recognize when it is called via the name `vimdiff`. It feels weird to suggest undoing a decision that was clearly made intentionally.

**Comment by bfredl** [comment]
>  I still don't know why it was decided that nvim should not recognize when it is called via the name vimdiff

because neovim is not vim. If we would it would rather be `nvimdiff` and our rationale is that if we change it, we could just as well change it to `nvim -d` form. most environments accepting a user command should allow command with params.

**Comment by RichardBronosky** [comment]
Anyway, for what it is worth, this is what I was thinking.

```
bin_dir="/usr/local/bin"
removed=(
    "vimdiff:nvim -d"
    "view:nvim -R"
    "ex:nvim -e"
    "exim:nvim -E"
    "rview:nvim -RZ"
    "rvim:nvim -Z"
)
for wrapper in "${removed[@]}"; do
    script="${bin_dir}/${wrapper%:*}"
    echo "${wrapper#*:}" '"$@"' >"$script"
    chmod +x "$script"
done
```

It works to recreate all the "aliases" listed on [the page][1]. But unlike aliases, it works outside of the shell.

[1]: https://neovim.io/doc/user/vim_diff.html#nvim-features-removed

**Comment by RichardBronosky** [comment]
> because neovim is not vim. If we would it would rather be `nvimdiff` and our rationale is that if we change it, we could just as well change it to `nvim -d` form. most environments accepting a user command should allow command with params.

I'm not saying that you should create the symlinks. I'm questioning why it was decided to **not** recognize when nvim is called by another name. If you decided to make it recognize when it was called by another name, and decided to require `nvimdiff` instead of `vimdiff` that would be horrible. I didn't start this thread to suggest that a horrible decision had been made. I just wanted to know why the decision was made.

But now that I can't even communicate the difference between an **alias** and a **symlink** or an **argument** (like `$1`) and recognizing the **name** called (like `$0`)… I don't know where to go from here. Maybe instead updating the documentation *which is all I was asking for **permission** to do*, this thread will be found by someone who needs it. ??

**Comment by bfredl** [comment]
>  If you decided to make it recognize when it was called by another name, and decided to require nvimdiff instead of vimdiff that would be horrible. I didn't start this thread to suggest that a horrible decision had been made.

I didn't claim you did suggest it either. I just answered your direct question _why_ this decision was made.

**Comment by RichardBronosky** [comment]
> I just answered your direct question _why_ this decision was made.

Understood. I just wanted to clarify the non-adversarial intent of my post. These communications are difficult because of the combined affects of (a) the dryness of text, (b) trend toward hostility on the internet, and (c) me having Asperger’s.

**Comment by dundargoc** [comment]
I played archeologist and found [this](https://github.com/neovim/neovim/issues/1646) issue to be the most relevant as to why vimdiff was closed. It's a pretty long issue, but in summary:

1. Using `vimdiff` would conflict with vim's `vimdiff`
2. It's redundant (and potentially confusing) having two ways of doing the same thing
3. There should be one and only one neovim. No nvim-tiny, nvim-huge, nvimdiff, envim, nvimnox...

Having extended reasoning for each decision made is an admirable goal, but I don't think this is reasonable (writing good documentation is *hard* and takes a lot of time), nor do I think it fits in the documentation.

I hope this answers you question.
